apisix/cli/etcd.lua

```diff
@@ -146,6 +146,241 @@ local function request(url, yaml_conf)
+        end
+
+        local res, err
+        local retry_time = 0
+        while retry_time < 2 do
+            res, err = request({
+                url = put_url,
+                method = "POST",
+                source = ltn12.source.string(post_json),
+                sink = ltn12.sink.table(response_body),
+                headers = headers
+            }, yaml_conf)
+            retry_time = retry_time + 1
+            if res then
+                break
+            end
+            print(str_format("Warning! Request etcd endpoint \'%s\' error, %s, retry time=%s",
+                                put_url, err, retry_time))
+        end
+
+        if not res then
+            errmsg = str_format("request etcd endpoint \"%s\" error, %s\n", put_url, err)
+            util.die(errmsg)
+        end
+
+        local res_put = table_concat(response_body)
+        if res_put:find("404 page not found", 1, true) then
+            errmsg = str_format("gRPC gateway is not enabled in etcd cluster \"%s\",",
+                                "which is required by Apache APISIX\n")
+            util.die(errmsg)
+        end
+
+        if res_put:find("CommonName of client sending a request against gateway", 1, true) then
+            errmsg = str_format("etcd \"client-cert-auth\" cannot be used with gRPC-gateway, "
+                                .. "please configure the etcd username and password "
+                                .. "in configuration file\n")
+            util.die(errmsg)
+        end
+
+        if res_put:find("error", 1, true) then
+            is_success = false
+            if (index == host_count) then
+                errmsg = str_format("got malformed key-put message: \"%s\" from etcd \"%s\"\n",
+                                    res_put, put_url)
+                util.die(errmsg)
+            end
+
+            break
+        end
+
+        if args and args["verbose"] then
+            print(res_put)
+        end
+    end
+
+    return is_success
+end
+
+
+local function grpc_request(url, yaml_conf, key)
+    local cmd
+
+    local auth = ""
+    if yaml_conf.etcd.user then
+        local user = yaml_conf.etcd.user
+        local password = yaml_conf.etcd.password
+        auth = str_format("--user=%s:%s", user, password)
+    end
+
+    if str_sub(url, 1, 8) == "https://" then
+        local host = url:sub(9)
+
+        local verify = true
+        local certificate, pkey, cafile
+        if yaml_conf.etcd.tls then
+            local cfg = yaml_conf.etcd.tls
+
+            if cfg.verify == false then
+                verify = false
+            end
+
+            certificate = cfg.cert
+            pkey = cfg.key
+
+            local apisix_ssl = yaml_conf.apisix.ssl
+            if apisix_ssl and apisix_ssl.ssl_trusted_certificate then
+                cafile = apisix_ssl.ssl_trusted_certificate
+            end
+        end
+
+        cmd = str_format(
+            "etcdctl --insecure-transport=false %s %s %s %s " ..
```

Comment:
So, `etcdctl` will be a dependency for APISIX to run?

Reply:
Yes. I will package it this week.